src/main/bootstrap/cache_validator.rs

```diff
@@ -0,0 +1,198 @@
+        let _ = core.insert_state(token, state.clone());
+
+        Ok(())
+    }
+
+    /// Sends STUN requests to cached peers that recently been inactive.
+    fn ping_inactive_peers(
+        &mut self,
+        core: &mut EventLoopCore,
+        poll: &Poll,
+        inactive_peers: HashSet<PeerInfo>,
+    ) {
+        debug!("Testing {} inactive cached peers.", inactive_peers.len());
+        for peer in inactive_peers {
+            self.ping_peer_and_cache_alive(core, poll, peer);
+        }
+    }
+
+    /// Send STUN request to given peer and wait for response.
+    /// If response arrives within timeout, peer is added to the bootstrap cache head.
+    fn ping_peer_and_cache_alive(&mut self, core: &mut EventLoopCore, poll: &Poll, peer: PeerInfo) {
+        let self_weak = self.self_weak.clone();
+        let expired_tx = self.expired_tx.clone();
+        let on_done = move |core: &mut EventLoopCore,
+                            _poll: &Poll,
+                            request_token,
+                            req_status: Result<SocketAddr, ()>| {
+            if let Some(self_rc) = self_weak.upgrade() {
+                let _ = self_rc.borrow_mut().sent_requests.remove(&request_token);
+                if req_status.is_ok() {
+                    let expired_peers = core.user_data_mut().bootstrap_cache.put(peer);
+                    if !expired_peers.is_empty() {
+                        let _ = expired_tx.send(expired_peers);
+                    }
+                }
+            }
+        };
+        if let Ok(request_token) = GetExtAddr::start(
+            core,
+            poll,
+            ipv4_addr(0, 0, 0, 0, 0),
+            &peer,
+            self.our_pk,
+            &self.our_sk,
+            Some(PEER_TEST_TIMEOUT_SEC),
+            Box::new(on_done),
+        ) {
+            let _ = self.sent_requests.insert(request_token);
+        }
+    }
+
+    /// Terminates any in-progress requests.
+    fn terminate_requests(&mut self, core: &mut EventLoopCore, poll: &Poll) {
+        for req_token in self.sent_requests.drain() {
+            let req_state = match core.get_state(req_token) {
+                Some(state) => state,
+                None => continue,
+            };
+            req_state.borrow_mut().terminate(core, poll);
+        }
+    }
+
+    fn read_expired_rx(&mut self, core: &mut EventLoopCore, poll: &Poll) {
+        while let Ok(expired_peers) = self.expired_rx.try_recv() {
+            self.ping_inactive_peers(core, poll, expired_peers);
+        }
+    }
+}
+
+impl State<CrustData> for CacheValidator {
+    fn ready(&mut self, core: &mut EventLoopCore, poll: &Poll, kind: Ready) {
+        if kind.is_readable() {
+            self.read_expired_rx(core, poll)
+        }
+    }
+
+    fn terminate(&mut self, core: &mut EventLoopCore, poll: &Poll) {
+        self.terminate_requests(core, poll);
+        let _ = core.remove_state(self.token);
```

Comment:
missing de-registeration of the `expired_rx` which remains in the poll-list as an orphan (i.e., without any associate `State` with it).